### PR TITLE
[WIP][SPARK-40143][SQL] ANSI mode: allow explicitly casting fraction strings as Integral types

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Currently,  for the following query it behaves differently between ansi mode and non-ansi mode:
```
-- ANSI on
> select cast('1.2' as int) 
CAST_INVALID_INPUT ERROR
```
```
-- ANSI off
> select cast('1.2' as int) 
1
```

The change was introduced in https://github.com/apache/spark/commit/ac262cb27255f989f6a6dd864bd5114a928b96da. With the change, we can avoid the following query returning true under ANSI mode
```
-- with ANSI on, both '1.2' and 1 will be cast as long type
select '1.2' = 1
```

However, the fraction strings should be allowed as per the SQL standard
image
image

This PR is to change the syntax for explicit casting to allow converting fraction strings as integral types. If the cast is implicit(not an explicit cast from SQL query), then it remains unchanged.

### Why are the changes needed?

To be more compliant with the SQL standard. Also, explicit casting string as integral types is now more consistent between the ANSI mode and the non-ANSI mode.

### Does this PR introduce _any_ user-facing change?

Yes, it allows explicitly casting fraction strings as Integral types in ANSI mode.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @srielau as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -509,6 +509,10 @@ case class Cast(
     evalMode == EvalMode.TRY
   }
 
+  private def isUserSpecified: Boolean = {
+    getTagValue(Cast.USER_SPECIFIED_CAST).getOrElse(false)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:513]: We shouldn't rely on tree node tag to define the behavior, as tree node tag is unstable today, and we only use it for non-important use cases, such as pretty print and error message.

**Comment by cloud-fan** [comment]
when adding implicit case for string to int, can we add a `AssertIntegral` expression? so the final expression will be `Cast(AssertIntegral(str_col), ...)`.

**Comment by cloud-fan** [comment]
I think we do have a narrow wrist?
```
  override def implicitCast(e: Expression, expectedType: AbstractDataType): Option[Expression] = {
    implicitCast(e.dataType, expectedType).map { dt =>
      if (dt == e.dataType) e else Cast(e, dt)
    }
  }
```

**Comment by gengliangwang** [comment]
@cloud-fan it doesn't cover all the type coercion rules

**Comment by cloud-fan** [comment]
We need a narrow wrist anyway, otherwise we need to change many type coercion rules to pass a new parameter to cast.

**Comment by gengliangwang** [comment]
It's hard to define which cast(string as int) syntax outside the explicit cast and implicit cast. 
Given our current Cast implementation is complicated enough, I will close this one for now.

**Pull request closed without merging**